### PR TITLE
Don't fail when rendering an `Unbounded` layout of a `group`ed Line

### LINKED ISSUE #91: Unbounded layout of `group`ed Line fails
```haskell
> layoutUnbounded = layoutPretty defaultLayoutOptions { layoutPageWidth = Unbounded }
> layoutUnbounded $ group Line
SFail
```

### PR DESCRIPTION
**Pull request by sjakobi**
~~Fixes~~ **Doesn't fix** #91.

---

TODO:

* [ ] Add a more complex test case
* [ ] Check whether any documentation needs updating or clarification.

### CODE REVIEW

File: prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs [hunk 1]
```diff
@@ -1782,6 +1790,7 @@ layoutWadlerLeijen
         -> SimpleDocStream ann -- ^ Choice B.
         -> SimpleDocStream ann -- ^ Choice A if it fits, otherwise B.
     selectNicer (FittingPredicate fits) lineIndent currentColumn x y
+      | isFail x = y
```
Comment (sjakobi) [inline prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs]: I was wondering whether we need to worry about `SFail` not appearing at the beginning of a `Union`. At least `fuse` doesn't seem to "hide" the `Fail` in a  `(Union Fail x)` though:

https://github.com/quchen/prettyprinter/blob/4d51362d7ea8bca0fe9e7ffec2ccd8fcd0065c0d/prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs#L1407

Comment (quchen) [reply prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs]: `isFail` is `False` if the document fails, but that’s the same thing the two fitting predicates (for `layoutPretty` and `layoutSmart`) yield. What does the explicit `isFail` clause change?

Comment (sjakobi) [reply prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs]: I've explained this in https://github.com/quchen/prettyprinter/issues/91#issuecomment-549033715.

The short version is that both fitting predicates shortcut to `True` when the `maxWidth` is `Nothing` – which is the case with `Unbounded` layout options.

**Comment by sjakobi** [comment]
This doesn't seem to fix the issue completely:

If I change `dhall`'s `layoutOpts` to use `Unbounded` and run its test suite with this branch, I still see a bunch of these errors:

> Exception: »SFail« must not appear in a rendered »SimpleDocStream«. This is a bug in the layout algorithm! Please report this as a bug

**Comment by sjakobi** [comment]
Maybe we should instead remove the shortcutting from the fitting predicates…

**Comment by sjakobi** [comment]
For reference, these are some of of the test failures:

```
      ./tests/format/letNewlineComments:                                                                                       FAIL
      ./tests/format/escapeSingleQuotedOpenInterpolation:                                                                      FAIL
      ./tests/format/issue1413:                                                                                                FAIL
      ./tests/format/importAccess:                                                                          
      ./tests/format/letComments:                                                                                              FAIL
      ./tests/format/innerMultiline:                                                                                           FAIL
    Issue #126:                                                                                                                FAIL
  tutorial
    Interpolation
      Example #1:                                                                                                              FAIL

**Comment by sjakobi** [comment]
Actually I wonder whether the current design with `Unbounded` as alternative to `AvailablePageWidth` makes sense at all. The entire difference between `layoutSmart` and `layoutPretty` lies in their fitting predicates – and with `Unbounded` you completely turn them off! `layoutSmart (LayoutOptions Unbounded)` and `layoutPretty (LayoutOptions Unbounded)` will always produce the same layout!

IMHO it would be better to have a separate function `layoutUnbounded` that handles `Union`s simply by checking whether the first alternative results in `SFail`.

**Comment by sjakobi** [comment]
Superseded by #97.

**Pull request closed without merging**